File: Problems/648. Replace Words/replace-words.java
//Solution 1: 16ms 55.34 MB
class Solution {
  class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isEndOfWord = false;
  }

  class Trie {
    TrieNode root = new TrieNode();

    public void insert(String word) {
      TrieNode node = root;
      for (char c : word.toCharArray()) {
        if (!node.children.containsKey(c)) {
          node.children.put(c, new TrieNode());
        }
        node = node.children.get(c);
      }
      node.isEndOfWord = true;
    }

    public String findPrefix(String word) {
      TrieNode node = root;
      StringBuilder prefix = new StringBuilder();

      for (char c : word.toCharArray()) {
        if (!node.children.containsKey(c)) {
          break;
        }
        node = node.children.get(c);
        prefix.append(c);
        if (node.isEndOfWord) {
          return prefix.toString();
        }
      }

      return "";
    }
  }

  public String replaceWords(List<String> dictionary, String sentence) {
    Trie trie = new Trie();

    for (String word : dictionary) {
      trie.insert(word);
    }

    String[] words = sentence.split(" ");
    StringBuilder result = new StringBuilder();

    for (String word : words) {
      String prefix = trie.findPrefix(word);
      if (!prefix.isEmpty()) {
        result.append(prefix).append(" ");
      } else {
        result.append(word).append(" ");
      }
    }

    return result.toString().trim();
  }
}

//Solution 2: 10ms 55.40 MB
class Solution {
  class TrieNode {
    TrieNode[] children = new TrieNode[26];
    boolean isEndOfWord = false;
  }

  class Trie {
    TrieNode root = new TrieNode();

    public void insert(String word) {
      TrieNode node = root;
      for (char c : word.toCharArray()) {
        int key = c - 'a';
        if (node.children[key] == null) {
          node.children[key] = new TrieNode();
        }
        node = node.children[key];
      }
      node.isEndOfWord = true;
    }

    public String findPrefix(String word) {
      TrieNode node = root;
      StringBuilder prefix = new StringBuilder();

      for (char c : word.toCharArray()) {
        int key = c - 'a';
        if (node.children[key] == null) {
          break;
        }
        node = node.children[key];
        prefix.append(c);
        if (node.isEndOfWord) {
          return prefix.toString();
        }
      }

      return "";
    }
  }

  public String replaceWords(List<String> dictionary, String sentence) {
    Trie trie = new Trie();

    for (String word : dictionary) {
      trie.insert(word);
    }

    String[] words = sentence.split(" ");
    StringBuilder result = new StringBuilder();

    for (String word : words) {
      String prefix = trie.findPrefix(word);
      if (!prefix.isEmpty()) {
        result.append(prefix).append(" ");
      } else {
        result.append(word).append(" ");
      }
    }

    return result.toString().trim();
  }
}

File: Problems/648. Replace Words/replace-words.ts
//Solution 1: Time Limit Exceed
function replaceWords(dictionary: string[], sentence: string): string {
  const dictionarySet = new Set(dictionary);
  const words = sentence.split(" ");
  let result = "";

  for (const word of words) {
    let prefix = "";
    for (let i = 0; i < word.length; i++) {
      prefix += word[i];
      if (dictionarySet.has(prefix)) {
        result += prefix + " ";
        prefix = "";
        break;
      }
    }

    if (prefix !== "") {
      result += word + " ";
    }
  }

  return result.trimEnd();
}

//Solution 2: 110ms 75.91 MB
class TrieNode {
  children: TrieNode[] = Array(26);
  isEndOfWord: boolean = false;
}

class Trie {
  root: TrieNode = new TrieNode();

  insert(word: string) {
    let node = this.root;
    for (const char of word) {
      const index = char.charCodeAt(0) - "a".charCodeAt(0);
      if (!node.children[index]) {
        node.children[index] = new TrieNode();
      }
      node = node.children[index];
    }
    node.isEndOfWord = true;
  }

  findPrefix(word: string): string {
    let node = this.root;
    let prefix = "";
    for (const char of word) {
      const index = char.charCodeAt(0) - "a".charCodeAt(0);
      if (!node.children[index]) break;
      node = node.children[index];
      prefix += char;
      if (node.isEndOfWord) return prefix;
    }
    return "";
  }
}

function replaceWords(dictionary: string[], sentence: string): string {
  const trie = new Trie();
  for (const word of dictionary) {
    trie.insert(word);
  }

  return sentence
    .split(" ")
    .map((word) => {
      const prefix = trie.findPrefix(word);
      return prefix || word;
    })
    .join(" ");
}

//Solution 3: 95ms 70.96 MB
class TrieNode {
  children: { [key: string]: TrieNode } = {};
  isEndOfWord: boolean = false;
}

class Trie {
  root: TrieNode = new TrieNode();

  insert(word: string) {
    let node = this.root;
    for (const char of word) {
      if (!node.children[char]) {
        node.children[char] = new TrieNode();
      }
      node = node.children[char];
    }
    node.isEndOfWord = true;
  }

  findPrefix(word: string): string {
    let node = this.root;
    let prefix = "";
    for (const char of word) {
      if (!node.children[char]) break;
      node = node.children[char];
      prefix += char;
      if (node.isEndOfWord) return prefix;
    }
    return "";
  }
}

function replaceWords(dictionary: string[], sentence: string): string {
  const trie = new Trie();
  for (const word of dictionary) {
    trie.insert(word);
  }

  return sentence
    .split(" ")
    .map((word) => {
      const prefix = trie.findPrefix(word);
      return prefix || word;
    })
    .join(" ");
}
